pkg/controller/gpudevice/gpu_controller.go

```diff
@@ -0,0 +1,255 @@
+	if gpu == nil || gpu.DeletionTimestamp != nil || gpu.Spec.NodeName != h.nodeName {
+		return gpu, nil
+	}
+
+	enabled, gpuStatus, err := gpuhelper.GenerateGPUStatus(filepath.Join(v1beta1.SysDevRoot, gpu.Spec.Address), gpu.Spec.NodeName)
+	if err != nil {
+		return gpu, fmt.Errorf("error generating status for SRIOVGPUDevice %s: %v", gpu.Name, err)
+	}
+
+	// perform enable/disable operation as needed
+	if gpu.Spec.Enabled != enabled {
+		logrus.Debugf("performing gpu management for %s", gpu.Name)
+		return h.manageGPU(gpu)
+	}
+
+	if !reflect.DeepEqual(gpu.Status, gpuStatus) {
+		logrus.Debugf("updating gpu status for %s:", gpu.Name)
+		gpu.Status = *gpuStatus
+		return h.sriovGPUClient.UpdateStatus(gpu)
+	}
+	return nil, nil
+}
+
+// SetupSRIOVGPUDevices is called by the node controller to reconcile objects on startup and predefined intervals
+func (h *Handler) SetupSRIOVGPUDevices() error {
+	sriovGPUDevices, err := gpuhelper.IdentifySRIOVGPU(h.options, h.nodeName)
+	if err != nil {
+		return err
+	}
+	return h.reconcileSRIOVGPUSetup(sriovGPUDevices)
+}
+
+// reconcileSRIOVGPUSetup runs the core logic to reconcile the k8s view of node with actual state on the node
+func (h *Handler) reconcileSRIOVGPUSetup(sriovGPUDevices []*v1beta1.SRIOVGPUDevice) error {
+	// create missing SRIOVGPUdevices, skipping GPU's which are already passed through as PCIDevices
+	for _, v := range sriovGPUDevices {
+		// if pcideviceclaim already exists for SRIOVGPU, then likely this GPU is already passed through
+		// skip creation of SriovGPUDevice object until PCIDeviceClaim exists
+		existingClaim, err := h.pciDeviceClaimCache.Get(v.Name)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return fmt.Errorf("error looking up pcideviceclaim for sriovGPUDevice %s: %v", v.Name, err)
+		}
+		if existingClaim != nil {
+			// pciDeviceClaim exists skipping
+			logrus.Debugf("skipping creation of vGPUDevice %s as PCIDeviceClaim exists", existingClaim.Name)
+			continue
+		}
+
+		if err := h.createOrUpdateSRIOVGPUDevice(v); err != nil {
+			return err
+		}
+	}
+	set := map[string]string{
+		v1beta1.NodeKeyName: h.nodeName,
+	}
+
+	existingGPUs, err := h.sriovGPUCache.List(labels.SelectorFromSet(set))
+	if err != nil {
+		return err
+	}
+
+	for _, v := range existingGPUs {
+		if !containsGPUDevices(v, sriovGPUDevices) {
+			if err := h.sriovGPUClient.Delete(v.Name, &metav1.DeleteOptions{}); err != nil {
+				return fmt.Errorf("error deleting non existent GPU device %s: %v", v.Name, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// createOrUpdateSRIOVGPUDevice will check and create GPU if one doesnt exist. If one is found it will perform an update if needed
+func (h *Handler) createOrUpdateSRIOVGPUDevice(gpu *v1beta1.SRIOVGPUDevice) error {
+	_, err := h.sriovGPUCache.Get(gpu.Name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			_, createErr := h.sriovGPUClient.Create(gpu)
+			return createErr
+		}
+	}
+
+	return err
+}
```

Comment:
Initial plan was to reconcile the real world and k8s api state accordingly. reconciling from host back into k8s api in initial versions led to issues post reboot. During a reboot the sriov devices are removed and gpu is unconfigured. This would result in gpu getting updated and vgpu devices being removed since gpu was now unconfigured. As a result the logic was removed to ensure we reconcile only from defined k8s api state to the host.